src/main/java/com/google/devtools/build/lib/packages/LabelPrinter.java

```diff
@@ -0,0 +1,104 @@
+package com.google.devtools.build.lib.packages;
+
+import com.google.devtools.build.lib.cmdline.Label;
+import com.google.devtools.build.lib.cmdline.PackageIdentifier;
+import com.google.devtools.build.lib.cmdline.RepositoryMapping;
+import com.google.devtools.build.lib.packages.semantics.BuildLanguageOptions;
+import net.starlark.java.eval.Printer;
+import net.starlark.java.eval.StarlarkSemantics;
+
+/** Knows how to print labels consistently in various formats. */
+public interface LabelPrinter {
+  /**
+   * Creates a {@link LabelPrinter} that prints labels in the same way as the Starlark `str` method.
+   * This behavior is useful when matching labels against Starlark values, in particular in tools.
+   *
+   * <p>Do not use this method directly, call {@link
+   * com.google.devtools.build.lib.query2.common.CommonQueryOptions#getLabelPrinter(StarlarkSemantics,
+   * RepositoryMapping)} instead.
+   */
+  static LabelPrinter canonical(StarlarkSemantics starlarkSemantics) {
```

Comment:
I think I'd prefer to call this `starlark` or something (and in the same vein, call the other `displayForm`). `canonical` could sometimes output "apparent" labels (ie. with bzlmod off) and `apparent`, canonical labels (ie. when not visible to main repo).